// sqlite3orm/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var AutoUpgrader_1 = require("./AutoUpgrader");
exports.AutoUpgrader = AutoUpgrader_1.AutoUpgrader;
exports.UpgradeMode = AutoUpgrader_1.UpgradeMode;
var BaseDAO_1 = require("./BaseDAO");
exports.BaseDAO = BaseDAO_1.BaseDAO;
var core_1 = require("./core");
exports.SQL_DEFAULT_SCHEMA = core_1.SQL_DEFAULT_SCHEMA;
exports.SQL_MEMORY_DB_PRIVATE = core_1.SQL_MEMORY_DB_PRIVATE;
exports.SQL_MEMORY_DB_SHARED = core_1.SQL_MEMORY_DB_SHARED;
exports.SQL_OPEN_CREATE = core_1.SQL_OPEN_CREATE;
exports.SQL_OPEN_DEFAULT = core_1.SQL_OPEN_DEFAULT;
exports.SQL_OPEN_READONLY = core_1.SQL_OPEN_READONLY;
exports.SQL_OPEN_READWRITE = core_1.SQL_OPEN_READWRITE;
exports.SqlConnectionPool = core_1.SqlConnectionPool;
exports.SqlDatabase = core_1.SqlDatabase;
exports.SqlStatement = core_1.SqlStatement;
var dbcatalog_1 = require("./dbcatalog");
exports.DbCatalogDAO = dbcatalog_1.DbCatalogDAO;
var metadata_1 = require("./metadata");
exports.field = metadata_1.field;
exports.Field = metadata_1.Field;
exports.fk = metadata_1.fk;
exports.FKDefinition = metadata_1.FKDefinition;
exports.getModelMetadata = metadata_1.getModelMetadata;
exports.id = metadata_1.id;
exports.IDXDefinition = metadata_1.IDXDefinition;
exports.index = metadata_1.index;
exports.METADATA_MODEL_KEY = metadata_1.METADATA_MODEL_KEY;
exports.MetaModel = metadata_1.MetaModel;
exports.MetaProperty = metadata_1.MetaProperty;
exports.PropertyType = metadata_1.PropertyType;
exports.Schema = metadata_1.Schema;
exports.schema = metadata_1.schema;
exports.table = metadata_1.table;
exports.Table = metadata_1.Table;
var query_1 = require("./query");
exports.QueryModel = query_1.QueryModel;
exports.TABLEALIAS = query_1.TABLEALIAS;
__export(require("./utils"));
